scipy/integrate/_tanhsinh.py

```diff
@@ -853,18 +882,29 @@ def _tanhsinh_iv(f, a, b, log, maxfun, maxlevel, minlevel,
         raise ValueError('`callback` must be callable.')
 
     return (f, a, b, log, maxfun, maxlevel, minlevel,
-            atol, rtol, args, preserve_shape, callback)
+            atol, rtol, args, preserve_shape, callback, xp)
 
 
-def _logsumexp(x, axis=0):
-    # logsumexp raises with empty array
-    x = np.asarray(x)
+def _logsumexp(x, /, *, axis=0, keepdims=False, xp=None):
+    # Dispatcher for _logsumexp until scipy/scipy#21149 merges
+    # Also, logsumexp raises with empty array; don't do that.
+    xp = array_namespace(x) if xp is None else xp
+    x = xp.asarray(x)
     shape = list(x.shape)
     if shape[axis] == 0:
         shape.pop(axis)
-        return np.full(shape, fill_value=-np.inf, dtype=x.dtype)
-    else:
-        return special.logsumexp(x, axis=axis)
+        return xp.full(shape, fill_value=-xp.inf, dtype=x.dtype)
+
+    if is_numpy(xp):
+        return special.logsumexp(x, axis=axis, keepdims=keepdims)
+    elif is_cupy(xp):
+        from cupyx.scipy.special import logsumexp
```

Comment:
To appease MyPy:
```suggestion
        from cupyx.scipy.special import logsumexp  # type: ignore[import-not-found]
```
The other error is `Instruction clone failed in 00:10` on Cirrus.